test/Tests/src/Semantic/Hashcode_Spec.enso

```diff
@@ -0,0 +1,34 @@
+from Standard.Test import Test
+
+type My_Type
+    Value foo bar baz
+
+spec =
+    Test.specify "Hashcode and equality contract" <|
+        # TODO: Feed various objects
+        obj1 = ...
+        obj2 = ...
+        if obj1 == obj2 then obj1.hash_code . should_equal obj2.hash_code
+        if obj1.hash_code == obj2.hash_code then obj1 . should_equal obj2
+        if obj1 != obj2 then obj1.hash_code . should_not_equal obj2.hash_code
+        if obj1.hash_code != obj2.hash_code then obj1 . should_not_equal obj2
+
+    Test.specify "Hashcodes are consistent" <|
+        old_hash = obj.hash_code
+        new_hash = obj.hash_code
+        old_hash . should_equal new_hash
+
+    Test.specify "Different hashcodes for different atoms" <|
+        (My_Type.Value 1 2 "blah").hash_code . should_not_equal (My_Type.Value 2 1 "blah").hash_code
+        ([1,2]).hash_code . should_not_equal ([2,1]).hash_code
+        ([1,2]).hash_code . should_not_equal ([1,1]).hash_code
+        ([1,2,3]).hash_code . should_not_equal ([2,1]).hash_code
+        ([1,2,3]).hash_code . should_not_equal ([1,2]).hash_code
+        ([1,2]).hash_code . should_equal ([1,2]).hash_code
+        ([1]).hash_code . should_not_equal 1.hash_code
+        1.hash_code . should_equal 1.hash_code
+
+    Test.specify "Hash codes of interop values" <|
+
```

Comment:
We need to ensure that the _Enso_ hashcode is consistent with our numeric and Text equality, this will require some tweaks as to how hashcodes are computed for numbers and texts:

```suggestion
2.hash_code . should_equal 2.0.hash_code
0.hash_code . should_equal 0.0.hash_code

# This should also work in atoms
(My_Type.Value -100 23.0 0).hash_code . should_equal (My_Type.Value -100.0 23 0).hash_code
(My_Type.Value -100 23.1 0).hash_code . should_not_equal (My_Type.Value -100.0 23 0).hash_code

# Text equality relies on the NFD normalization of Unicode strings, so:
'ś'.hash_code . should_equal 's\u0301'.hash_code
'e\u0301abc'.hash_code . should_equal 'éabc'.hash_code
'e\u0301abc'.hash_code . should_not_equal 'é'.hash_code
```

Reply:
Unfortunately if we want to keep the above, I'm afraid it may not be enough to rely on `identityHashCode` as we will need custom hashcode logic for the primitive types - numbers and strings.

Unless this requirement is changing, but that would probably require changing the behaviour of `==` too and I don't think we are planning to go this route?

Reply:
Sounds good, I just wanted to make sure we can handle this properly. In particular that  this is also correctly handled for Atom fields.

I see `hashCodes[i] = fieldsInterop.identityHashCode(field);` suggesting that we will delegate to `identityHashCode` for fields, but it seems that we need something a bit more complex there to handle our overrides for the builtin types.

Note that we should definitely _not_ convert _all_ doubles to integers - this should only happen for doubles that satisfy `x % 1.0 == 0.0`. We do _not_ want `1.5.hash_code == 1.0.hash_code`.

As for texts, please note that currently `Text.==` is implemented in `Standard.Base` (IIRC) and relies on `Text_Utils.equals`. If we move the NFD normalization to a builtin hashcode method, we should probably move `Text.==` to be a builtin too - so that we can more easily ensure both equality and hashcode of text are kept in sync.

Alternatively (IMO it may be even a better solution if it does not come with performance costs), we can keep the current design and since `Text.==` is a custom override in our stdlib instead of a builtin, we could also do an analogous override for `Text.hash_code`.

My point is - I don't care too much about where `Text.==` and `Text.hash_code` are implemented (stdlib seems a bit better to me as there are many other text methods there so it may make sense to keep them in one place, but that is just my opinion and I'm not even very attached to it), but let's try to keep them both in a single place (either builtins or stdlib).

Reply:
I will definitelly add these tests for the primitive types in the final testing.

Reply:
> I see hashCodes[i] = fieldsInterop.identityHashCode(field); suggesting that we will delegate to identityHashCode for fields, but it seems that we need something a bit more complex there to handle our overrides for the builtin types.

Yes, you are right. But still, I consider this rather an implementation detail. In other words, I believe that handling custom hash codes for primitive types should not be a problem. `Atom.computeIdentityHashCode` is just a suggestion, and what we can do instead is that `Atom.identityHashCode` can forward the hashcode computation for its fields to `GetHashCodeNode`, which would have specializations for primitives.

As for your other comments about numbers and Text handling, I shall bear that in mind once I move forward with this PR.

Reply:
Adding to the care needed about `2 == 2.0` in Enso, reposted from Discord, so we do not loose track of this:

> the problem I'm potentially seeing is that if you pass Enso primitives 2 and 2.0 to Java, while in Enso they are `==`, in Java they are not. So I'm not sure if we should encourage creating Java hashmaps keyed by Enso objects without a special notion of equality, as then someone may key by these and get results inconsistent with Enso.

IMHO this is yet another suggestion that we should consider writing an Enso-Java Polyglot Handbook with best practices.
I'm not sure if there are any other ways of warning users that they need to be extra careful when creating Enso-centric Hashmaps in Java.
One additional thing that we may provide is a way to normalize the Enso objects to some common form that will be handled by Java correctly. For example, currently we are using `foldNumeric` in `UnorderedMultiValueKey` to fold `Double 2.0` to a `Long 2`.  We could expose that as a library (maybe part of `Polyglot_Utils` in `base`) to be used by any users - but we still need to make sure users are aware of its existence to use it.